transaction.go

```diff
@@ -0,0 +1,489 @@
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s *FactoidTransaction) TotalFCTInputs() (total uint64) {
+	return FactoidTransactionIOs(s.FCTInputs).TotalAmount()
+}
+
+func (s *FactoidTransaction) TotalFCTOutputs() (total uint64) {
+	return FactoidTransactionIOs(s.FCTOutputs).TotalAmount()
+}
+
+// TotalECOutput is delimated in factoishis
+func (s *FactoidTransaction) TotalECOutput() (total uint64) {
+	return FactoidTransactionIOs(s.ECOutputs).TotalAmount()
+}
+
+func (s FactoidTransactionIOs) TotalAmount() (total uint64) {
+	for _, io := range s {
+		total += io.Amount
+	}
+	return
+}
+
+func (s FactoidTransactionSignature) Validate(msg Bytes) bool {
+	return s.ReedeemCondition.Validate(msg, s.SignatureBlock)
+}
+
+// Get queries factomd for the entry corresponding to f.TransactionID, which must be not
+// nil. After a successful call all inputs, outputs, and the header will be populated
+func (f *FactoidTransaction) Get(c *Client) error {
+	// TODO: Test this functionality
+	// If the TransactionID is nil then we have nothing to query for.
+	if f.TransactionID == nil {
+		return fmt.Errorf("txid is nil")
+	}
+	// If the Transaction is already populated then there is nothing to do. If
+	// the Hash is nil, we cannot populate it anyway.
+	if f.IsPopulated() {
+		return nil
+	}
+	requestTxId := f.TransactionID
+
+	params := struct {
+		Hash *Bytes32 `json:"hash"`
+	}{Hash: f.TransactionID}
+	var result struct {
+		Data Bytes `json:"data"`
+	}
+	if err := c.FactomdRequest("raw-data", params, &result); err != nil {
+		return err
+	}
+
+	if err := f.UnmarshalBinary(result.Data); err != nil {
+		return err
+	}
+
+	txid, err := f.ComputeTransactionID()
+	if err != nil {
+		return err
+	}
+	if txid != *requestTxId {
+		return fmt.Errorf("invalid txid")
+	}
+
+	return nil
+}
+
+// ComputeTransactionID computes the txid for a given transaction. The txid is the sha256 of
+// the ledger fields in a factoid transaction. The ledger fields exclude the signature block of
+// the transaction
+func (f *FactoidTransaction) ComputeTransactionID() (Bytes32, error) {
+	data, err := f.MarshalLedgerBinary()
+	if err != nil {
+		return Bytes32{}, err
+	}
+
+	txid := Bytes32(sha256.Sum256(data))
+	return txid, nil
+}
+
+// ComputeFullHash computes the fullhash for a given transaction. The fullhash is the sha256 of all
+// the fields in a factoid transaction.
+func (f *FactoidTransaction) ComputeFullHash() (*Bytes32, error) {
+	data, err := f.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	txid := Bytes32(sha256.Sum256(data))
+	return &txid, nil
+}
+
+// MarshalLedgerBinary marshals the transaction ledger fields to their binary representation.
+// This excludes the signature blocks
+func (f *FactoidTransaction) MarshalLedgerBinary() ([]byte, error) {
+	// TODO: More checks up front?
+	if !f.IsPopulated() {
+		return nil, fmt.Errorf("not populated")
+	}
+
+	// It's very difficult to know the size before marshaling, as each in/out has a varint
```

Comment:
Check out my new varintf package API. I added a BufLen() function which determines the exact length of a varintf encoding for a given uint64. I also added a Put function to allow the user to control allocations.

Reply:
Each input/output in the body has a varint. To compute the size up front is not trivial.